### DISCUSSION TITLE
Custom metrics and dependencies are written twice to AppInsights by Azure Functions

### DISCUSSION THREAD

**Discussion by fgheysels** [arcus-azure/arcus.observability#193] [Q&A]
I use Arcus.Observability in an Azure Function.
I typically have a configuration-setting in my Azure Function which is named `APPINSIGHTS_INSTRUMENTATIONKEY` to [integrate my Azure Function with App Insights](https://docs.microsoft.com/en-us/azure/azure-functions/functions-monitoring#application-insights-integration).   This gives me built-in logging of my Function's usage to appinsights.

However, when I log custom metrics and dependencies, those metrics and dependencies are written twice to Application Insights:  once as a metric (or a dependency) which is as expected, but they're also written as a 'trace' with a warning level:

![image](https://user-images.githubusercontent.com/3605786/110533151-fb97db00-811d-11eb-936c-afc0fd6571b5.png)

Could it have something to do with [this](https://github.com/arcus-azure/arcus.observability/discussions/189) ?

I can offcourse change the name of the `APPINSIGHTS_INSTRUMENTATIONKEY` setting, but then I loose some functionality (like logs and metrics from the function host, so I rather don't do that.

**Comment by tomkerkhove** [comment]
@stijnmoreels Can you list all the providers that are added by Azure Functions please?

As far as I know, I think we cannot avoid this and the only way to mitigate this is use a different name for your key, but let's verify!

**Reply by stijnmoreels** [reply]
> @stijnmoreels Can you list all the providers that are added by Azure Functions please?

When using our Serilog setup, in an Azure Functions project, before we call this `ClearLoggerProviders...` method, there are two `ILoggerProvider` types registered:
* `Microsoft.Azure.WebJobs.Script.Diagnostics.HostFileLoggerProvider`
* `Microsoft.Azure.WebJobs.Script.Diagnostics.FunctionFileLoggerProvider`
Both are the ones that we filter on in the `ClearLoggerProviders...` method.
